node/p2p_block_sync.md

```diff
@@ -0,0 +1,84 @@
+# P2P Block Sync
+
+## Abstract
+
+P2P Block Sync enables rollkit full nodes, including aggregators, to gossip blocks amongst  
+themselves and have rollup lower block times than the DA layer block times.
+
+```mermaid
+sequenceDiagram
+    title P2P Block Sync
+
+    participant User
+    participant Block Producer
+    participant Full Node 1
+    participant Full Node 2
+    participant DA Layer
+
+    User->>Block Producer: Send Tx
+    Block Producer->>Block Producer: Generate Block
+    Block Producer->>DA Layer: Publish Block
+
+    Block Producer->>Full Node 1: Gossip Block
+    Block Producer->>Full Node 2: Gossip Block
+    Full Node 1->>Full Node 1: Verify Block
+    Full Node 1->>Full Node 2: Gossip Block
+    Full Node 1->>Full Node 1: Mark Block Soft-Confirmed
+
+    Full Node 2->>Full Node 2: Verify Block
+    Full Node 2->>Full Node 2: Mark Block Soft-Confirmed
+
+    DA Layer->>Full Node 1: Retrieve Block
+    Full Node 1->>Full Node 1: Mark Block Hard-Confirmed
+
+    DA Layer->>Full Node 2: Retrieve Block
+    Full Node 2->>Full Node 2: Mark Block Hard-Confirmed
+```
+
+## Protocol/Component Description
+
+P2P Block Sync consists of the following components:
+
+* block exchange service: responsible for gossiping blocks over P2P
+* block publication to P2P network
+* block retrieval from P2P network
+
+### Block Exchange Service
+
+The block exchange service is created during full node initialization. After that, during the block manager's initialization, a pointer to the block store inside block exchange service is passed to it. Blocks created in the block manager which need to gossipped over a p2p network are then passed to the `BlockCh` channel and then sent to [go-header] service to be gossip blocks over P2P network.
+
+### Block Publication to P2P network
+
+Blocks ready to be published to the P2P network are sent to the `BlockCh` channel in Block Manager inside `publishLoop`.
+The `blockPublishLoop` in the full node continuously listens for new blocks from the `BlockCh` channel and when a new block  
+is received, it is written to the block store and broadcasted to the network using the block exchange service.
+
+### Block Retrieval from P2P network
+
+Blocks gossiped to validating full nodes through the P2P network are retreived from the `Block Store` in `BlockStoreRetrieveLoop` in Block Manager.
+For every `blockTime` unit of time, a signal is sent to the `blockStoreCh` channel in block manager and when this signal is received, the  
+`BlockStoreRetrieveLoop` retrieves blocks from the block store. It keeps track of the last retrieved block's height and if the current block store's height  is greater than the last retrieved block's height, it retrieves all blocks from the block store that are between these two heights.
```

Comment:
Is there a limit to how many heights you can retrieve?

Reply:
no limit, just all the blocks between the two heights described.

Reply:
That means if I sync and I am far behind, I would be able to sync fully no matter the height difference. I think we should test how the sync of a chain (with a lot of txs) performs. Can be a follow-up.

Reply:
Yes exactly you should be able to do that if you're far behind. We should test it.